From ce96aa2731d93425f4d9584a888ad42d41466a3e Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Wed, 11 Mar 2026 14:20:44 +0800
Subject: [PATCH] mtk: mt76: mt7996: clear ieee80211_tx_info->status before use

Because the ieee80211_tx_info->status shares memory with other members
that belong to the same union, it must be initialized before being used
to prevent residual data from previous usage
(e.g., ieee80211_tx_info->control is use in the TX path).

However, mt76 uses ieee80211_tx_info->status->status_driver_data across
the TX and TxS path, so the field cannot be cleared.

Do the initialization in the following cases,
1. In _mt76_tx_complete_skb(), if the cb->pktid < MT_PACKET_ID_FIRST,
   the upper layer does not need the TxS, so we just clear the
   ieee80211_tx_info->status to prevent the garbage data from being used.
2. In __mt76_tx_status_skb_done(), if the skb's TxS and TxFree are both
   ready, we clear the ieee80211_tx_info->status before using it.

Do NOT do the initialization in the following case
1. In mt76_tx_status_skb_get(), which calls __mt76_tx_status_skb_done()
   with input flags MT_TX_CB_TXS_FAILED and MT_TX_CB_TXS_DONE being on.
   The TxS of the skb is timed out and will not be sent to the upper layer.

issue severity: High
Link: https://github.com/mediatek/mtk-openwrt-feeds/commit/4b573a7e757e53e638521be26d686084df058bf2
Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 tx.c | 7 +++++++
 1 file changed, 7 insertions(+)

Index: mt76-2026.09.01~be5ce791/tx.c
===================================================================
--- mt76-2026.09.01~be5ce791.orig/tx.c
+++ mt76-2026.09.01~be5ce791/tx.c
@@ -99,6 +99,9 @@ __mt76_tx_status_skb_done(struct mt76_de
 	if ((flags & done) != done)
 		return;
 
+	memset(&info->status, 0, sizeof(info->status) -
+				 sizeof(*info->status.status_driver_data));
+
 	/* Tx status can be unreliable. if it fails, mark the frame as ACKed */
 	if (flags & MT_TX_CB_TXS_FAILED &&
 	    (dev->drv->drv_flags & MT_DRV_IGNORE_TXS_FAILED)) {
@@ -300,6 +303,10 @@ void __mt76_tx_complete_skb(struct mt76_
 
 	if (cb->pktid < MT_PACKET_ID_FIRST) {
 		struct ieee80211_rate_status rs = {};
+		struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
+
+		memset(&info->status, 0, sizeof(info->status) -
+					 sizeof(*info->status.status_driver_data));
 
 		hw = mt76_tx_status_get_hw(dev, skb);
 		status.sta = wcid_to_sta(wcid);
